fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,673 @@
+        root for root in store.blocks.keys()
+        if store.blocks[root].parent_root == block_root
+    ]
+
+    if any(children):
+        # Get leaves of all children and add to the set.
+        leaves = set().union(*[get_leaf_block_roots(store, child) for child in children])
+        return leaves
+    else:
+        # This block is a leaf.
+        return set([block_root])
+
+```
+
+#### `get_current_epoch_participating_indices`
+
+```python
+def get_epoch_participating_indices(
+    state: BeaconState, 
+    active_validator_indices: Sequence[ValidatorIndex],
+    current_epoch: bool
+) -> Set[ValidatorIndex]: 
+    if current_epoch:
+        epoch_participation = state.current_epoch_participation
+    else:
+        epoch_participation = state.previous_epoch_participation
+
+    return set([
+        i for i in active_validator_indices 
+        if has_flag(epoch_participation[i], TIMELY_TARGET_FLAG_INDEX)
+    ])
+```
+
+#### `get_ffg_support`
+
+```python
+def get_ffg_support(store: Store, block_root: Root) -> Gwei:
+    """
+    Returns the total weight supporting the checkpoint in the block's chain at block's epoch
+    """
+
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+    current_epoch = get_current_store_epoch(store)
+
+    assert current_epoch <= block_epoch + 1
+    assert block_epoch <= current_epoch
+
+    current_epoch = get_current_store_epoch(store)
+    checkpoint_root = get_checkpoint_block(store, block_root, block_epoch)
+    checkpoint = Checkpoint(root=checkpoint_root, epoch=block_epoch)
+
+    if checkpoint not in store.checkpoint_states:
+        return Gwei(0)
+
+    checkpoint_state = store.checkpoint_states[checkpoint]
+
+    leave_roots = get_leaf_block_roots(store, block_root)
+
+    active_validator_indices = get_active_validator_indices(checkpoint_state, block_epoch)
+    participating_indices = set().union(*[
+        get_epoch_participating_indices(
+            store.block_states[root], 
+            active_validator_indices, 
+            block_epoch == current_epoch
+        )
+        for root in leave_roots
+    ])
+
+    return get_total_balance(checkpoint_state, participating_indices)
+```
+
+
+
+#### `is_ffg_confirmed_current_epoch`
+
+```python
+def is_ffg_confirmed_current_epoch(
+    total_active_balance: int,
+    ffg_support_for_checkpoint: int,
+    remaining_ffg_weight: int,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,   
+) -> bool:
+    """
+    Returns whether the branch will justify it's current epoch checkpoint at the end of this epoch
+    """
+
+    max_adversarial_ffg_support_for_checkpoint = int(
+        min(
+            (total_active_balance * confirmation_byzantine_threshold - 1) // 100 + 1,
+            confirmation_slashing_threshold,
+            ffg_support_for_checkpoint
+        )
+    )
+
+    """
+    Returns whether the following condition is true using only integer arithmetic
```

Comment:
Suggest putting this as an additional note in the function's docstring

Reply:
This comment uses variables defined at the beginning of the method so I am afraid that moving it to the function's docstring would confuse readers.

I have left the comment where it was but replaced the docstring comments with standard in-line comments.